### ISSUE TITLE
If set `type: module` package.json, ts path-alias not replace, and can not resolve alias path module.

### ISSUE THREAD

**Issue by Innei** [nestjs/nest#10267]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

Can't replace alias path if file in nested-folder, but in root dir such as `/src` replace correctly.

```ts
import { Module } from '@nestjs/common';
import { AppController } from '~/app.controller.js'; // right
import { AppService } from '~/app.service.js'; // right
import { TestModule } from '~/test/index.module.js'; // wrong
```

transform to:

```ts
import { Module } from '@nestjs/common';
import { AppController } from "./app.controller.js";
import { AppService } from "./app.service.js";
import { TestModule } from '~/test/index.module.js'; // not replaced!!!
```

### Minimum reproduction code

https://stackblitz.com/edit/nestjs-typescript-starter-pnhuer?file=dist/app.module.js&terminal=start

### Steps to reproduce

> npm start:dev

```

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '~' imported from /home/projects/nestjs-typescript-starter-pnhuer/dist/app.module.js
    at InternalError.get (https://nestjs-typescript-starter-pnhuer.w.staticblitz.com/blitz.a1e315a41ec6498e2c2758b32fbf437db67b244b.js:6:292488)
    at process.emit (/home/projects/nestjs-typescript-starter-pnhuer/node_modules/source-map-support/source-map-support.js:508:52)
    at process.onGlobalUncaughtException [as _fatalException] (https://nestjs-typescript-starter-pnhuer.w.staticblitz.com/blitz.a1e315a41ec6498e2c2758b32fbf437db67b244b.js:6:251442)
    at Object.triggerUncaughtException (https://nestjs-typescript-starter-pnhuer.w.staticblitz.com/blitz.a1e315a41ec6498e2c2758b32fbf437db67b244b.js:15:356815)
    at i.loadESM (https://nestjs-typescript-starter-pnhuer.w.staticblitz.com/blitz.a1e315a41ec6498e2c2758b32fbf437db67b244b.js:6:246694)
    at handleMainPromise (https://nestjs-typescript-starter-pnhuer.w.staticblitz.com/blitz.a1e315a41ec6498e2c2758b32fbf437db67b244b.js:6:989050)

```

### Expected behavior

replace tsc alias path.

### Package

- [ ] I don't know. Or some 3rd-party package
- [ ] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [X] Other (see below)

### Other package

@nestjs/cli

### NestJS version

9.0.11

### Packages versions

```json
 "@nestjs/cli": "9.1.2",
```

### Node.js version

v16.14.2

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [X] Linux

**Comment by jmcdo29** [comment]
Looks like `tsconfig-paths` (what Nest is using under the hood for replacing path aliases) doesn't support ESM. [There's a number of open issues in their repo about it.](https://github.com/dividab/tsconfig-paths/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+esm)

**Comment by Cristiandi** [comment]
so what it means is, when `tsconfig-paths` support ESM nestjs will be able to work with ESM with no problems? is that right?
_hehe sorry for the silly question_ 😅

**Comment by jmcdo29** [comment]
> so what it means is, when `tsconfig-paths` support ESM nestjs will be able to work with ESM with no problems? is that right? _hehe sorry for the silly question_ sweat_smile

Partially, there's still more to it than just `tsconfig-paths`. There's ensuring we can still support express, fastify, apollo, mercurius, all the microservice transports and WS, and ensuring that unless we upgrade at a major version we don't introduce any breaking changes for CJS users as well. Not to mention testing tools supporting ESM and migrating the entire code base over to ESM as well. 

In short: it's something we're working on and monitoring, but it's gonna take a while

**Comment by Cristiandi** [comment]
Buffff that seems to be a huge n important, anyway thanks a lot for working on it and on this amazing framework, gonna be expecting the ESM compatibility changes.
Thanks for your explanation 💪🏻

**Comment by yuntian001** [comment]
I had the same problem. How did you solve it

**Comment by gterras** [comment]
> I had the same problem. How did you solve it

That's the neat thing about esm/cjs interop, you don't. Most likely your situation can be solved another way like changing an extension to `.mjs`. 

Worth noting that https://github.com/privatenumber/tsx is often mentioned in https://github.com/dividab/tsconfig-paths issues, it looks like a viable alternative.
